tinnsleep/create_reports.py

```diff
@@ -0,0 +1,149 @@
+                     fir_design='firwin', filter_length='auto', phase='zero-double',
+                     picks=picks_chan)
+
+    # Creating epochs
+    epochs = RawToEpochs_sliding(raw, duration=duration, interval=interval)
+
+    # Epoch rejection based on |min-max| thresholding
+    params = dict(ch_names=raw.info["ch_names"],
+                  rejection_thresholds=dict(emg=1e-04),  # two order of magnitude higher q0.01
+                  flat_thresholds=dict(emg=1e-09),  # one order of magnitude lower median
+                  channel_type_idx=dict(emg=[0, 1]),
+                  full_report=True
+                  )
+    amplitude_labels, bad_lists = is_good_epochs(epochs, **params)
+
+    suppressed_amp = np.sum(np.invert(amplitude_labels))
+
+    # Reuniting the rejection algorithms
+    valid_labels = valid_labels = np.invert(np.any(np.c_[impedance_labels, np.invert(amplitude_labels)], axis=-1))
+    suppressed_all = np.sum(np.invert(valid_labels))
+
+    if get_log:
+        # Creating log report
+        log = {"suppressed_imp_THR": suppressed_imp, "suppressed_amp_THR": suppressed_amp,
+                                         "suppressed_overall": suppressed_all}
+
+        return epochs, valid_labels, log
+    else:
+        return epochs, valid_labels
+
+
+def reporting(epochs, valid_labels, THR_classif, log={}):
+    """creates clinical reports of bruxism out of a epoch array, for different thresholding values
+       Parameters
+       ----------
+       epochs : ndarray, shape (n_epochs, n_channels, duration)
+            Epoched view of `data`. Epochs are in the first dimension.
+       valid_labels : lsit of booleans
+            labels of the epochs as good (True) or bad (False) for future annotation and reporting
+       THR_classif : list of couple of floats
+           list of couples of absolute and relative thresholds values of the classifier to test
+       log : OPTIONAL dictionary
+            logs of the preprocessing steps, including the number of epochs rejected at each step
+
+       Returns
+       -------
+
+        dictionary
+           Dictionary containing fields:
+           - THR_classif: threshold(s) of classification tested
+           - labels : list of labels of bursts for the epochs of the recording for each THR_classif in the same order
+           - reports: list of clinical reports of the recording for each THR_classif tested in the same order
+           - log : log of the pre-processing operations
+       """
+    labs = []
+    reps = []
+    # for each value of THR_classif, create a report and a list of labels
+    for THR in THR_classif:
+        pipeline = AmplitudeThresholding(abs_threshold=THR[0], rel_threshold=THR[1])
+        X = rms(epochs[valid_labels])  # take only valid labels
+        labels = pipeline.fit_predict(X)
+        print(labels)
+        report = generate_clinical_report(labels)
+        labels = fuse_with_classif_result(np.invert(valid_labels),
```

Comment:
are you sure you want to `generate_clinical_report` before `fuse_with_classif_result` ?

Reply:
Good point, I thought it was a good idea but I don't remember why, will check tomorrow

Reply:
@RobinGuillard ?

Reply:
I thought about it and it is essential to leave it that way. In fact, when we do the "fuse_with_classif_result" step, we add "False" values for all epochs considered as bad, which lengthens the data without adding more events. Consequently, if we would have run the report after fuse_with_classif_result, we would have had wrong values for several indexes of the report, that are depending on the length of the recording and of the clean data. 
We should definitely let it in this order.
Nonetheless, one drawback of the current report I'm making is that I don't save the initial length of the recording. I made a small modification that helps: I added a new field in the "log" file: I added the field "Initial_nb_epochs" for the total number of epochs before the preprocessing steps.